pint/pandas_interface/pint_array.py

```diff
@@ -0,0 +1,609 @@
+        self._data = self._coerce_to_pint_array(values, dtype=dtype, copy=copy)
+
+    def _coerce_to_pint_array(self, values, dtype=None, copy=False):
+        if isinstance(values, self._dtype.type):
+            return values
+
+        if is_list_like(values):
+            if all(is_bool(v) for v in values):
+                # known bug in pint https://github.com/hgrecco/pint/issues/673
+                raise TypeError("Invalid magnitude for {}: {}"
+                                "".format(self._dtype.type, values))
+
+            for i, v in enumerate(values):
+                if isinstance(v, self._dtype.type):
+                    continue
+                else:
+                    values[i] = v * self._find_first_unit(values)
+
+            units = set(v.units for v in values)
+            if len(units) > 1:
+                # need to work out a way to test this
+                raise TypeError("The units of all quantities are not the same"
+                                " for input {}".format(values))
+
+            magnitudes = [v.magnitude for v in values]
+
+            return self._dtype.type(magnitudes, values[0].units)
+
+        return NotImplementedError
+
+    def _find_first_unit(self, values):
+        for v in values:
+            if isinstance(v, self._dtype.type):
+                return v.units
+
+        return self._dtype.type(1).units
+
+    def __getitem__(self, item):
+        # type (Any) -> Any
+        """Select a subset of self.
+        Parameters
+        ----------
+        item : int, slice, or ndarray
+            * int: The position in 'self' to get.
+            * slice: A slice object, where 'start', 'stop', and 'step' are
+              integers or None
+            * ndarray: A 1-d boolean NumPy ndarray the same length as 'self'
+        Returns
+        -------
+        item : scalar or PintArray
+        """
+        if is_integer(item):
+            return self._data[item]
+
+        return type(self)(self._data[item])
+
+    def __len__(self):
+        # type: () -> int
+        """Length of this array
+
+        Returns
+        -------
+        length : int
+        """
+        return len(self._data)
+
+    def __repr__(self):
+        """
+        Return a string representation for this object.
+
+        Invoked by unicode(df) in py2 only. Yields a Unicode String in both
+        py2/py3.
+        """
+
+        klass = self.__class__.__name__
+        data = format_object_summary(self, default_pprint, False)
+        attrs = format_object_attrs(self)
+        space = " "
+
+        prepr = (u(",%s") %
+                 space).join(u("%s=%s") % (k, v) for k, v in attrs)
+
+        res = u("%s(%s%s)") % (klass, data, prepr)
+
+        return res
+
+    def __array__(self, dtype=None, copy=False):
+    # this is necessary for some pandas operations, eg transpose
+    # note, units will be lost
+        if dtype == None:
+            dtype=object
+        if type(dtype) == str:
```

Comment:
why (here and below) are you using `dtype ==` instead of isinstance

Reply:
Also, pint defines `compat.string_types` just in case you need to quicklly check for string and unicode in Py2 and str in Py3

Reply:
fixed